src/arch/src/x86_64/mod.rs

```diff
@@ -17,15 +19,41 @@ pub mod msr;
 pub mod regs;
 
 use linux_loader::configurator::linux::LinuxBootConfigurator;
+use linux_loader::configurator::pvh::PvhBootConfigurator;
 use linux_loader::configurator::{BootConfigurator, BootParams};
 use linux_loader::loader::bootparam::boot_params;
-use vm_memory::{Address, GuestAddress, GuestMemory, GuestMemoryMmap, GuestMemoryRegion};
-
+use linux_loader::loader::elf::start_info::{
+    hvm_memmap_table_entry, hvm_modlist_entry, hvm_start_info,
+};
+use vm_memory::{
+    Address, ByteValued, GuestAddress, GuestMemory, GuestMemoryMmap, GuestMemoryRegion,
+};
+
+use super::BootProtocol;
 use crate::InitrdConfig;
 
 // Value taken from https://elixir.bootlin.com/linux/v5.10.68/source/arch/x86/include/uapi/asm/e820.h#L31
 const E820_RAM: u32 = 1;
 
+// Workaround for the Rust orphan rules that guarantee trait coherence by wrapping the foreign type
```

Comment:
We worked with Alejandro ~2 years and a half ago when he first proposed these patches and we got them upstreamed as part of rust-vmm (https://github.com/rust-vmm/linux-loader/pull/11). Since Firecracker also consumes the crate I think it makes sense for these to be consumed from upstream rather than re-implementing them here. You can check out the example in doc.rs: https://docs.rs/linux-loader/0.6.0/linux_loader/

Reply:
I'm making use of all the changes I could find which went into linux-loader.  There's a reason why this PR is +495 -78 compared to Alejandro's original +1169 -93 PR!

Reply:
Oops there's a few lines here which didn't need to be here.  New patch coming shortly.

Reply:
Yes, there were a few things that were duplicated. Sorry about that, should've been more explicit.